### PR TITLE
Navigate to YT from Duck Player

### PR DESCRIPTION
**Pull request by CrisBarreiro**
Task/Issue URL: https://app.asana.com/0/0/1207562793310532/f 

### Steps to test this PR

_Feature 1_
- [ ]
- [ ]

### UI changes
| Before  | After |
| ------ | ----- |
!(Upload before screenshot)|(Upload after screenshot)|

### CODE REVIEW

**Comment by CrisBarreiro** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4780?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4780** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4780?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#4752** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4752?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4748** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4748?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4663** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4663?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @CrisBarreiro and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: app/src/main/java/com/duckduckgo/app/browser/WebViewRequestInterceptor.kt [hunk 1]
```diff
@@ -159,6 +122,10 @@ class WebViewRequestInterceptor(
             return WebResourceResponse(null, null, null)
         }
 
+        if (url != null) {
```
Comment (marcosholgado) [inline app/src/main/java/com/duckduckgo/app/browser/WebViewRequestInterceptor.kt:125]: Tests?

File: duckplayer/duckplayer-impl/src/main/java/com/duckduckgo/duckplayer/impl/RealDuckPlayer.kt [hunk 2]
```diff
@@ -17,31 +17,57 @@
...
+private const val DUCK_PLAYER_DOMAIN = "player"
+private const val DUCK_PLAYER_URL_BASE = "$duck://$DUCK_PLAYER_DOMAIN/"
+private const val DUCK_PLAYER_ASSETS_PATH = "duckplayer/"
+private const val DUCK_PLAYER_ASSETS_INDEX_PATH = "${DUCK_PLAYER_ASSETS_PATH}index.html"
+private const val REFERER_HEADER = "Referer"
+private const val REFERRING_QUERY_PARAM = "embeds_referring_euri"
 
+@SingleInstanceIn(AppScope::class)
 @ContributesBinding(AppScope::class)
 class RealDuckPlayer @Inject constructor(
```
Comment (marcosholgado) [inline duckplayer/duckplayer-impl/src/main/java/com/duckduckgo/duckplayer/impl/RealDuckPlayer.kt:59]: Tests?

**Comment by marcosholgado** [comment]
Had a go at the test cases and found these behaviours not matching them.

**Back button with Always Ask**
- Open a video in YouTube
- Go to Duck Player
- On the Duck Player page use the back button on your phone
- The video should load in YouTube with no overlay shown <- This doesn't happen for me and the overlay is always shown.

**Back button with Always**
- Open a video in YouTube, check Duck Player is loaded
- On the Duck Player page use the back button on your phone
- The video should load in YouTube <- This never happens and I'm taken back to the new tab screen instead. Which I think makes sense because YT never actually loads.

**Back button age restricted videos**
- Open video using Duck Player URL
- On the Duck Player page use the back button on your phone
- Check YouTube is loaded <- as above, YT doesn't load, instead I'm taken to the new tab screen.

**Comment by CrisBarreiro** [comment]
> Had a go at the test cases and found these behaviours not matching them.
> 
> **Back button with Always Ask**
> 
>     * Open a video in YouTube
> 
>     * Go to Duck Player
> 
>     * On the Duck Player page use the back button on your phone
> 
>     * The video should load in YouTube with no overlay shown <- This doesn't happen for me and the overlay is always shown.
> 
> 
> **Back button with Always**
> 
>     * Open a video in YouTube, check Duck Player is loaded
> 
>     * On the Duck Player page use the back button on your phone
> 
>     * The video should load in YouTube <- This never happens and I'm taken back to the new tab screen instead. Which I think makes sense because YT never actually loads.
> 
> 
> **Back button age restricted videos**
> 
>     * Open video using Duck Player URL
> 
>     * On the Duck Player page use the back button on your phone
> 
>     * Check YouTube is loaded <- as above, YT doesn't load, instead I'm taken to the new tab screen.

@marcosholgado thanks for having a look!

Back button with Always Ask -> I think the current behavior makes sense, as the page the user had seen before contains the overlay, and we probably shouldn't treat it in the same way as tapping the YouTube button

For the other 2 cases, as you mentioned, since YouTube never actually loads, it doesn't make sense to navigate back to it. That was a mistake on my site when writing the test cases

File: duckplayer/duckplayer-impl/src/main/java/com/duckduckgo/duckplayer/impl/RealDuckPlayer.kt (hunk 2 above)
Comment (CrisBarreiro) [reply duckplayer/duckplayer-impl/src/main/java/com/duckduckgo/duckplayer/impl/RealDuckPlayer.kt:59]: Tested in the top of the stack

File: app/src/main/java/com/duckduckgo/app/browser/WebViewRequestInterceptor.kt (hunk 1 above)
Comment (CrisBarreiro) [reply app/src/main/java/com/duckduckgo/app/browser/WebViewRequestInterceptor.kt:125]: Added to top of the stack

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt [hunk 3]
```diff
@@ -1092,7 +1092,7 @@ class BrowserTabViewModel @Inject constructor(
         when (stateChange) {
             is WebNavigationStateChange.NewPage -> {
                 val uri = stateChange.url.toUri()
-                if (duckPlayer.isYoutubeNoCookie(uri)) {
+                if (duckPlayer.isSimulatedYoutubeNoCookie(uri)) {
```
[the remark this answers is not in the document]
Comment (CrisBarreiro) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:1095]: Updated in top of the stack

File: app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt [hunk 4]
```diff
@@ -383,12 +383,15 @@ class BrowserWebViewClient @Inject constructor(
                         pageLoadedHandler.onPageLoaded(it, navigationList.currentItem?.title, safeStart, currentTimeProvider.elapsedRealtime())
                         shouldSendPagePaintedPixel(webView = webView, url = it)
                         appCoroutineScope.launch(dispatcherProvider.io()) {
-                            if (duckPlayer.isYoutubeNoCookie(url)) {
+                            if (duckPlayer.isSimulatedYoutubeNoCookie(url)) {
```
[the remark this answers is not in the document]
Comment (CrisBarreiro) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt:386]: Updated in top of the stack

**Review by marcosholgado — APPROVED** [review]
LGTM

**Pull request merged**